### PR TITLE
fixes #29: Implement ForOfStatement

### LINKED ISSUE #29: ES6: for-of
# [13.6](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-iteration-statements) Iteration Statements

### Syntax

<pre>
IterationStatement[Yield, Return] :
  do Statement[?Yield, ?Return] while ( Expression[In, ?Yield] ) ;opt
  while ( Expression[In, ?Yield] ) Statement[?Yield, ?Return]
  for ( [lookahead ∉ {let [}] Expression[?Yield]opt ; Expression[In, ?Yield]opt ; Expression[In, ?Yield]opt ) Statement[?Yield, ?Return]
  for ( var VariableDeclarationList[?Yield] ; Expression[In, ?Yield]opt ; Expression[In, ?Yield]opt ) Statement[?Yield, ?Return]
  for ( LexicalDeclaration[?Yield] Expression[In, ?Yield]opt ; Expression[In, ?Yield]opt ) Statement[?Yield, ?Return]
  for ( [lookahead ∉ {let [}] LeftHandSideExpression[?Yield] in Expression[In, ?Yield] ) Statement[?Yield, ?Return]
  for ( var ForBinding[?Yield] in Expression[In, ?Yield] ) Statement[?Yield, ?Return]
  for ( ForDeclaration[?Yield] in Expression[In, ?Yield] ) Statement[?Yield, ?Return]
  for ( [lookahead ≠ let ] LeftHandSideExpression[?Yield] of AssignmentExpression[In, ?Yield] ) Statement [?Yield, ?Return]
  for ( var ForBinding[?Yield] of AssignmentExpression[In, ?Yield] ) Statement[?Yield, ?Return]
  for ( ForDeclaration[?Yield] of AssignmentExpression[In, ?Yield] ) Statement[?Yield, ?Return]

ForDeclaration[Yield] :
  LetOrConst ForBinding[?Yield]

ForBinding[Yield] :
  BindingIdentifier[?Yield]
  BindingPattern[?Yield]
</pre>

### Static Semantics: Early Errors

- It is a Syntax Error if IsLabelledFunction(Statement) is true for any occurrence of Statement in these rules.

# [13.6.4](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-for-in-and-for-of-statements) The for-in and for-of Statements

### Static Semantics: Early Errors

```
IterationStatement :
  for ( LeftHandSideExpression in Expression ) Statement
  for ( LeftHandSideExpression of AssignmentExpression ) Statement
```

- It is a Syntax Error if LeftHandSideExpression is either an ObjectLiteral or an ArrayLiteral and if the lexical token sequence matched by LeftHandSideExpression cannot be parsed with no tokens left over using AssignmentPattern as the goal symbol.

If LeftHandSideExpression is either an ObjectLiteral or an ArrayLiteral and if the lexical token sequence matched by LeftHandSideExpression can be parsed with no tokens left over using AssignmentPattern as the goal symbol then the following rules are not applied. Instead, the Early Error rules for AssignmentPattern are used.

- It is a Syntax Error if IsValidSimpleAssignmentTarget of LeftHandSideExpression is false.

- It is a Syntax Error if the LeftHandSideExpression is CoverParenthesizedExpressionAndArrowParameterList : ( Expression ) and Expression derives a production that would produce a Syntax Error according to these rules if that production is substituted for LeftHandSideExpression. This rule is recursively applied.

**NOTE** The last rule means that the other rules are applied even if parentheses surround Expression.

```
IterationStatement :
  for ( ForDeclaration in Expression ) Statement
  for ( ForDeclaration of AssignmentExpression ) Statement
```

- It is a Syntax Error if the BoundNames of ForDeclaration contains "let".

- It is a Syntax Error if any element of the BoundNames of ForDeclaration also occurs in the VarDeclaredNames of Statement.

- It is a Syntax Error if the BoundNames of ForDeclaration contains any duplicate entries.

### PR DESCRIPTION
**Pull request by benvinegar**
Fixes #29

### CODE REVIEW

File: src/parser.js [hunk 1]
```diff
@@ -607,10 +607,14 @@ export class Parser extends Tokenizer {
         let initDecl = this.parseVariableDeclaration();
         this.allowIn = previousAllowIn;
 
-        if (initDecl.declarators.length === 1 && this.match(TokenType.IN)) {
+        if (initDecl.declarators.length === 1 && this.match(TokenType.IN) || this.match(TokenType.OF)) {
```
Comment (michaelficarra) [inline src/parser.js]:
```js
if (initDecl.declarators.length === 1 && (this.match(TokenType.IN) || this.match(TokenType.OF))) {
```

File: src/parser.js [hunk 2]
```diff
@@ -607,10 +607,14 @@ export class Parser extends Tokenizer {
...
+          let type = this.lookahead.type;
```
Comment (michaelficarra) [inline src/parser.js]:
```js
let type = this.lookahead.type === TokenType.IN ? Shift.ForInStatement : Shift.ForOfStatement;
```

File: src/parser.js [hunk 3]
```diff
@@ -607,10 +607,14 @@ export class Parser extends Tokenizer {
...
+
           this.lex();
           right = this.parseExpression();
-          return new Shift.ForInStatement(initDecl, right, this.getIteratorStatementEpilogue());
+          return type === TokenType.IN ?
```
Comment (michaelficarra) [inline src/parser.js]:
```js
return new type(initDecl, right, this.getIteratorStatementEpilogue());
```

File: src/parser.js (hunk 1 above)
Comment (benvinegar) [inline src/parser.js]: Whoops, had this line but must have accidentally deleted during some cleanup.

File: src/parser.js [hunk 4]
```diff
@@ -607,10 +607,13 @@ export class Parser extends Tokenizer {
...
 
-        if (initDecl.declarators.length === 1 && this.match(TokenType.IN)) {
+        if (initDecl.declarators.length === 1 && (this.match(TokenType.IN) || this.match(TokenType.OF))) {
+          let type = this.match(TokenType.IN) ?
+            Shift.ForInStatement : Shift.ForOfStatement;
+
           this.lex();
           right = this.parseExpression();
-          return new Shift.ForInStatement(initDecl, right, this.getIteratorStatementEpilogue());
+          return type(initDecl, right, this.getIteratorStatementEpilogue());
```
Comment (michaelficarra) [inline src/parser.js]: `new type`

**Comment by michaelficarra** [comment]
Good enough for now. We'll add more tests later.

**Pull request merged**